packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1459 @@
+     *         exchange.
+     *
+     * @param validator Address of the validator.
+     * @param tokenId   The tokenId of KGH to delegate.
+     *
+     * @return The amount of KRO shares that the Vault would exchange for the KGH provided.
+     * @return The amount of KGH shares that the Vault would exchange for the KGH provided.
+     */
+    function delegateKgh(
+        address validator,
+        uint256 tokenId
+    ) external checkIsActive(validator) returns (uint128, uint128) {
+        uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenId);
+        uint128 kroShares = previewDelegate(validator, kroInKgh);
+        uint128 kghShares = previewKghDelegate(validator);
+        _delegateKgh(validator, tokenId, kroInKgh, kroShares, kghShares);
+
+        emit KghDelegated(validator, msg.sender, tokenId, kroInKgh, kroShares, kghShares);
+        return (kroShares, kghShares);
+    }
+
+    /**
+     * @notice Delegate KGHs to the validator and returns the amount of shares that the vault would
+     *         exchange.
+     *
+     * @param validator Address of the validator.
+     * @param tokenIds  The tokenIds of KGHs to delegate.
+     *
+     * @return The amount of KRO shares that the Vault would exchange for the KGHs provided.
+     * @return The amount of KGH shares that the Vault would exchange for the KGHs provided.
+     */
+    function delegateKghBatch(
+        address validator,
+        uint256[] calldata tokenIds
+    ) external checkIsActive(validator) returns (uint128, uint128) {
+        uint128 kroShares;
+        uint128 kghShares = previewKghDelegate(validator);
+        uint128 kroInKghs;
+
+        for (uint256 i = 0; i < tokenIds.length; ) {
+            KGH.safeTransferFrom(msg.sender, address(this), tokenIds[i]);
+
+            uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenIds[i]);
+            uint128 kroSharesForTokenId = previewDelegate(validator, kroInKgh);
+
+            unchecked {
+                kroInKghs += kroInKgh;
+                kroShares += kroSharesForTokenId;
+
+                _vault[validator].kghDelegators[msg.sender].kroShares[
+                    tokenIds[i]
+                ] = kroSharesForTokenId;
+                _vault[validator].kghDelegators[msg.sender].kghShares[tokenIds[i]] = kghShares;
+
+                ++i;
+            }
+        }
+
+        kghShares *= uint128(tokenIds.length);
+        _delegateKghBatch(validator, uint128(tokenIds.length), kroInKghs, kroShares, kghShares);
+
+        emit KghBatchDelegated(validator, msg.sender, tokenIds, kroShares, kghShares);
+        return (kroShares, kghShares);
+    }
+
+    /**
+     * @notice Initiate KRO undelegation for given validator and per-mille.
+     *
+     * @param validator Address of the validator.
+     * @param perMille  The Per-mille of shares to undelegate.
+     */
+    function initUndelegate(address validator, uint128 perMille) external {
```

Comment:
Do you have any specific reason of using ratio for undelegation? 
IMHO, we can simply implement this with argument of the amount of share.

Reply:
That would make the implementation simpler, but I thought about the integration with frontend, and it would be better to have a precentage-based argument rather than the amount of share itself. How do you think about this?

Reply:
In that case, I think it would be better to use the amount of share token itself.
The way we need to prefer should be to make contract as simple as possible.

Reply:
Ok, then I'll address it. Thanks!